pkg/controller/direct/privilegedaccessmanager/entitlement_controller.go

```diff
@@ -0,0 +1,442 @@
+			"and 'spec.requesterJustificationConfig.unstructured' should be configured: neither is configured")
+	}
+	if obj.Spec.RequesterJustificationConfig.NotMandatory != nil && obj.Spec.RequesterJustificationConfig.Unstructured != nil {
+		return nil, fmt.Errorf("one and only one of 'spec.requesterJustificationConfig.notMandatory' " +
+			"and 'spec.requesterJustificationConfig.unstructured' should be configured: both configured")
+	}
+
+	// Get privilegedaccessmanager GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func oneOfContainer(ctx context.Context, reader client.Reader, obj *krm.PrivilegedAccessManagerEntitlement, projectRef *refs.ProjectRef, folderRef *refs.FolderRef, organizationRef *refs.OrganizationRef) (string, error) {
+	container := ""
+	if projectRef == nil && folderRef == nil && organizationRef == nil {
+		return "", fmt.Errorf("none of 'projectRef', 'folderRef' or 'organizationRef' was set")
+	}
+	if folderRef == nil && organizationRef == nil {
+		project, err := refs.ResolveProject(ctx, reader, obj, projectRef)
+		if err != nil {
+			return "", err
+		}
+		projectID := project.ProjectID
+		if projectID == "" {
+			return "", fmt.Errorf("cannot resolve project: project ID is empty")
+		}
+		container = fmt.Sprintf("projects/%s", projectID)
+	}
+	if projectRef == nil && organizationRef == nil {
+		folder, err := refs.ResolveFolder(ctx, reader, obj, folderRef)
+		if err != nil {
+			return "", err
+		}
+		folderID := folder.FolderID
+		if folderID == "" {
+			return "", fmt.Errorf("cannot resolve folder: folder ID is empty")
+		}
+		container = fmt.Sprintf("folders/%s", folderID)
+	}
+	if projectRef == nil && folderRef == nil {
+		organization, err := refs.ResolveOrganization(ctx, reader, obj, organizationRef)
+		if err != nil {
+			return "", err
+		}
+		organizationID := organization.OrganizationID
+		if organizationID == "" {
+			return "", fmt.Errorf("cannot resolve organization: organization ID is empty")
+		}
+		container = fmt.Sprintf("organizations/%s", organizationID)
+	}
+	if container == "" {
+		return "", fmt.Errorf("more than one of 'projectRef', 'folderRef' "+
+			"or 'organizationRef' was set: projectRef: %+v, folderRef: %+v, organizationRef: %+v",
+			projectRef, folderRef, organizationRef)
+	}
+	return container, nil
+}
+
+func (m *entitlementModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id        *PrivilegedAccessManagerEntitlementIdentity
+	gcpClient *gcp.Client
+	desired   *krm.PrivilegedAccessManagerEntitlement
+	actual    *privilegedaccessmanagerpb.Entitlement
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
```

Comment:
Nit: you probably don't need ctrlName, because I think we have the GVK anyway, and I think we've _also_ added  a controller name in the controller-runtime infrastructure.  Not a blocker!

Reply:
Are you suggesting we remove this line?

Reply:
Discussed offline. It's to remove `.WithName(ctrlName)`.